src/main/scala/workflow/AutoCacheRule.scala

```diff
@@ -0,0 +1,355 @@
+        case node: WeightedNode => (i, node.weight)
+        case _ => (i, 1)
+      }
+      case (_, i) => (i, 1)
+    }.toMap
+  }
+
+  /**
+   * Get a map representing the immediate children for each instruction
+   * Note: This doesn't capture how many times each child depended on the instruction
+   */
+  def getImmediateChildrenByInstruction(instructions: Seq[Instruction]): Map[Int, Seq[Int]] = {
+    instructions.indices.map(i => (i, WorkflowUtils.getImmediateChildren(i, instructions))).toMap
+  }
+
+  /**
+   * Get an estimate for how many times each instruction will be executed, assuming
+   * the given set of instructions have their outputs cached
+   */
+  def getRuns(instructions: Seq[Instruction], cache: Set[Int], nodeWeights: Map[Int, Int]): Map[Int, Int] = {
+    val immediateChildrenByInstruction = getImmediateChildrenByInstruction(instructions)
+
+    instructions.indices.foldRight(Map[Int, Int]()) { case (i, runsByIndex) =>
+      if (immediateChildrenByInstruction(i).isEmpty) {
+        runsByIndex + (i -> 1)
+      }
+      else {
+        val runs = immediateChildrenByInstruction(i).map(j => if (cache.contains(j)) {
+          nodeWeights(j)
+        } else {
+          nodeWeights(j) * runsByIndex(j)
+        }).sum
+        runsByIndex + (i -> runs)
+      }
+    }
+  }
+
+  def generalizeProfiles(newScale: Long, sampleProfiles: Seq[SampleProfile]): Profile = {
+    def getModel(inp: Iterable[(Long, String, Long)]): Double => Double = {
+      val observations = inp.toArray
+
+      //Pack a data matrix with observations
+      val X = DenseMatrix.ones[Double](observations.length, 2)
+      observations.zipWithIndex.foreach(o => X(o._2, 0) = o._1._1.toDouble)
+      val y = DenseVector(observations.map(_._3.toDouble))
+      val model = max(X \ y, 0.0)
+
+      //A function to apply the model.
+      def res(x: Double): Double = DenseVector(x, 1.0).t * model
+
+      res
+    }
+
+    val samples = sampleProfiles.flatMap { case SampleProfile(scale, value) =>
+      Array(
+        (scale, "memory", value.mem),
+        (scale, "time", value.ns)
+      )}.groupBy(a => a._2)
+
+    val models = samples.mapValues(getModel)
+
+    Profile(models("time").apply(newScale).toLong, models("memory").apply(newScale).toLong)
+  }
+
+  def profileInstructions(
+    instructions: Seq[Instruction],
+    scales: Seq[Long],
+    numTrials: Int
+  ): Map[Int, Profile] = {
+    val instructionsToProfile = instructions.indices.toSet -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions)
+
+    val registers = new Array[InstructionOutput](instructions.length)
+    val numPerPartitionPerNode = scala.collection.mutable.Map[Int, Map[Int, Int]]()
+    val profiles = scala.collection.mutable.Map[Int, Profile]()
+
+    val sortedScales = scales.sorted
+    for ((instruction, i) <- instructions.zipWithIndex if instructionsToProfile.contains(i)) {
+      instruction match {
```

Comment:
Hmm, if you think that the linear profile generalization would be sufficiently accurate then I can go ahead and add some code to enable this.